graphql-parser/src/source.rs

```diff
@@ -1,23 +1,46 @@
+use std::io;
+use std::io::prelude::*;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
 static DEFAULT_SOURCE_NAME: &'static str = "GraphQL";
 
 #[derive(Debug)]
-pub struct Source<'a> {
-  pub name: &'a str,
-  pub body: &'a str,
+pub struct Source {
+  pub name: String,
+  pub body: String,
 }
 
-impl<'a> Source<'a> {
-  pub fn new(body: &'a str) -> Source<'a> {
+impl Source {
+  pub fn new(body: String) -> Self {
     Source {
-      name: DEFAULT_SOURCE_NAME,
+      name: DEFAULT_SOURCE_NAME.to_owned(),
       body: body,
     }
   }
 
-  pub fn with_name(name: &'a str, body: &'a str) -> Source<'a> {
+  pub fn with_name(name: String, body: String) -> Self {
     Source {
       name: name,
       body: body,
     }
   }
-}
\ No newline at end of file
+
+  pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self, io::Error> {
+    let mut p = PathBuf::new();
+    p.push(path);
+    let mut f = try!(File::open(p.clone()));
+    let mut s = String::new();
+    try!(f.read_to_string(&mut s));
+    Ok(Source::with_name(Self::get_file_name(p), s))
+  }
+
+  fn get_file_name(path: PathBuf) -> String {
```

Comment:
This probably shouldn't be in the `impl` block, in fact this probably shouldn't be a function at all. Here it is with functional programming:

```rust
path.file_name().and_then(OsStr::to_str).unwrap_or(DEFAULT_SOURCE_NAME).to_owned()
```

And if that line is a little too long for your taste:

```rust
let name = {
  path
  .file_name()
  .and_then(OsStr::to_str)
  .unwrap_or(DEFAULT_SOURCE_NAME)
  .to_owned()
};
```

Much easier to understand then the `if let Some/Ok`s.